src/chocl8.zig

```diff
@@ -0,0 +1,303 @@
+    0xF0, 0x80, 0xF0, 0x80, 0xF0, // E
+    0xF0, 0x80, 0xF0, 0x80, 0x80, // F
+};
+
+/// instructions_per_cycle is the number of instructions per a cycle at 60Hz
+pub const instructions_per_cycle = 12;
+
+const Instruction = struct {
+    short: u16,
+    w: u4,
+    x: u4,
+    y: u4,
+    n: u4,
+    nn: u8,
+    nnn: u12,
+
+    fn decode(lo: u8, hi: u8) Instruction {
+        return Instruction{
+            .short = @as(u16, hi) << 8 | @as(u16, lo),
+            .w = @truncate(u4, hi >> 4),
+            .x = @truncate(u4, hi),
+            .y = @truncate(u4, lo >> 4),
+            .n = @truncate(u4, lo),
+            .nn = lo,
+            .nnn = @as(u12, hi & 0xF) << 8 | @as(u12, lo),
+        };
+    }
+};
+
+const Chip8 = struct {
+    /// mem is 4096 bytes of memory
+    /// 0x0 to 0x1FF is reserved for internal use
+    /// 0x200 is the entry point at which a ROM should be loaded
+    mem: [4096]u8 = [_]u8{0} ** 4096,
+    /// stack is a stack of size 16 for tracking mem locations
+    stack: [16]u16 = [_]u16{0} ** 16,
+    /// sp is the stack pointer
+    sp: u8 = 0,
+    // pc is the program counter
+    pc: u16 = 0,
+    /// index is a register for pointing at mem locations
+    index: u16 = 0,
+    /// delay is a counter decremented at 60hz
+    delay: u8 = 0,
+    /// sound is a counter decremented at 60hz and produces a tone when sound > 0
+    sound: u8 = 0,
+    /// reg is a 16 wide bank of 8-bit registers
+    reg: [16]u8 = [_]u8{0} ** 16,
+    // fbuf is the framebuffer of the screen
+    fbuf: [32]u64 = [_]u64{0} ** 32,
+    // counter tracks the number of instructions executed
+    counter: u64 = 0,
+
+    fn init() Chip8 {
+        var c8 = Chip8{};
+        //@memset(@ptrCast([*]u8, &c8), 0, @sizeOf(Chip8));
+        @memcpy(c8.mem[0x50..], font[0..], font.len);
+        return c8;
+    }
+
+    pub fn loadRom(self: *Chip8, data: []const u8) void {
+        // validate size will fit in memory
+        if (self.mem.len - 0x200 < data.len) {
+            // return error
+        }
+        @memcpy(self.mem[0x200..], @ptrCast([*]const u8, data[0..]), data.len);
+        self.pc = 0x200;
+    }
+
+    pub fn step(self: *Chip8, keys: u16) void {
+        const inst = Instruction.decode(self.mem[self.pc + 1], self.mem[self.pc]);
+        self.pc += 2;
+
+        self.counter += 1;
+        if (self.counter % instructions_per_cycle == 0) {
+            if (self.delay > 0) self.delay -= 1;
+            if (self.sound > 0) self.sound -= 1;
+        }
+
+        switch (inst.w) {
+            0x0 => switch (inst.nn) {
+                0xEE => {
+                    std.log.info("{}: ret", .{self.pc - 2});
+                    self.pc = self.stack[self.sp];
+                    self.sp -= 1;
+                },
+                0xE0 => {
+                    std.log.info("{}: clear fbuf", .{self.pc - 2});
+                    @memset(@ptrCast([*]u8, self.fbuf[0..]), 0, @sizeOf(u64) * self.fbuf.len);
```

Comment:
Same as `@memcpy`, `@memset` is an unsafe, low level intrinsic, and you should likely prefer to use `std.mem.set`.